src/artemis/stores/mapgraph/read.cljs

```diff
@@ -138,38 +138,72 @@
-              (if (:return-partial? gql-context)
-                result
-                (reduced ::incomplete-value)))
-
-            (map? expr)
-            (let [map-result (pull-join store result expr entity gql-context)]
-              (if (incomplete? map-result)
-                (reduced map-result)
-                map-result))
-
-            (= '* expr)                                     ; don't re-merge things we already joined
-            (merge result (apply dissoc entity (keys result)))
-
-            :else
-            (throw (ex-info "Invalid form in pull pattern"
-                            {:reason      ::invalid-pull-form
-                             ::form       expr
-                             ::pattern    pattern
-                             ::lookup-ref lookup-ref})))))
-      {}
-      pattern)))
+  (let [ref (:artemis.mapgraph/ref lookup-ref)]
+    (if (or (nil? ref) (contains? entities ref))
+      (when-let [entity (get entities (:artemis.mapgraph/ref lookup-ref))]
+        (reduce
+         (fn [result expr]
+           (let [{:keys [expr entity selection]} (expr-and-entity-for-gql expr entity gql-context)]
+             (cond
+               (keyword? expr)
+               (let [k (if (aliased? selection)
+                         (-> selection :name keyword)
+                         expr)]
+                 (if-let [[_ val] (find entity expr)]
+                   (assoc result k val)
+                   ;; Keyword was in query, but not found in store
+                   (if (:return-partial? gql-context)
+                     (assoc result k ::incomplete-value)
+                     (reduced ::incomplete-value))))
+
+               (map? expr)
+               (let [map-result (pull-join store result expr entity gql-context)]
+                 (if (incomplete? map-result)
+                   (reduced map-result)
+                   map-result))
+
+               (= '* expr)                                  ; don't re-merge things we already joined
+               (merge result (apply dissoc entity (keys result)))
+
+               :else
+               (throw (ex-info "Invalid form in pull pattern"
+                               {:reason      ::invalid-pull-form
+                                ::form       expr
+                                ::pattern    pattern
+                                ::lookup-ref lookup-ref})))))
+         {}
+         pattern))
+      ::incomplete-value)))
+
+(defn incomplete-or-has-incomplete? [val]
+  (or (incomplete? val)
+      (and (coll? val)
+           (not-empty (filter incomplete? val)))))
+
+(defn remove-incomplete [val]
+  (cond
+    (map? val)
+    (->> val
+         (remove (fn [[k v]] (incomplete-or-has-incomplete? v)))
+         (map (fn [[k v]] [k (remove-incomplete v)]))
+         (into {}))
+
+    (coll? val)
+    (like val (map remove-incomplete val))
+
+    :else
+    val))
+
+(defn has-incomplete? [val]
```

Comment:
this functions detects an `incomplete-value` from anywhere in the deep map. It might be a good idea to rename it to be more clear and/or add doc strings